crates/astria-composer/src/searcher/bundle_factory.rs

```diff
@@ -0,0 +1,479 @@
+    }
+
+    /// Replace self with a new empty bundle, returning the old bundle.
+    fn flush(&mut self) -> SizedBundle {
+        mem::replace(self, Self::new(self.max_size))
+    }
+
+    /// Consume self and return the underlying buffer of actions.
+    fn into_actions(self) -> Vec<Action> {
+        self.buffer
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum BundleFactoryError {
+    #[error("sequence action is larger than the max bundle size. seq_action size: {size}")]
+    SequenceActionTooLarge { size: usize, max_size: usize },
+}
+
+/// Manages the bundling of sequence actions into `SizedBundle`s. A `Vec<Action>` is flushed and
+/// added to the `finished` queue when an incoming `SequenceAction` won't fit in the current bundle.
+/// The `finished` queue operates in FIFO order, where `Vec<Action>`s are added to the back and
+/// taken off from the front.
+pub(super) struct BundleFactory {
+    /// The current bundle being built.
+    curr_bundle: SizedBundle,
+    /// The queue of bundles that have been built but not yet sent to the sequencer.
+    finished: VecDeque<SizedBundle>,
+}
+
+impl BundleFactory {
+    pub(super) fn new(max_bytes_per_bundle: usize) -> Self {
+        Self {
+            curr_bundle: SizedBundle::new(max_bytes_per_bundle),
+            finished: VecDeque::new(),
+        }
+    }
+
+    /// Buffer `seq_action` into the current bundle. If the bundle won't fit `seq_action`, flush
+    /// `curr_bundle` into the `finished` queue and start a new bundle
+    pub(super) fn try_push(
+        &mut self,
+        seq_action: SequenceAction,
+    ) -> Result<(), BundleFactoryError> {
+        let seq_action_size = estimate_size_of_sequence_action(&seq_action);
+
+        match self.curr_bundle.push(seq_action) {
+            Err(SizedBundleError::SequenceActionTooLarge(_seq_action)) => {
+                // reject the sequence action if it is larger than the max bundle size
+                Err(BundleFactoryError::SequenceActionTooLarge {
+                    size: seq_action_size,
+                    max_size: self.curr_bundle.max_size,
+                })
+            }
+            Err(SizedBundleError::NotEnoughSpace(seq_action)) => {
+                // if the bundle is full, flush it and start a new one
+                self.finished.push_back(self.curr_bundle.flush());
+                self.curr_bundle
+                    .push(seq_action)
+                    .expect("seq_action should not be larger than max bundle size");
+                Ok(())
+            }
+            Ok(()) => {
+                trace!(
+                    new_bundle_size = self.curr_bundle.curr_size,
+                    seq_action_size = seq_action_size,
+                    "bundled new sequence action"
+                );
+                Ok(())
+            }
+        }
+    }
+
+    /// Get the next bundle from the `finished` queue.
+    pub(super) fn next_finished(&mut self) -> Option<NextFinishedBundle> {
```

Comment:
why does this return `NextFinishedBundle` instead of just doing the `pop_front` directly?

Reply:
or rather, when/why is this used instead of `pop_now`?

Reply:
This is required to work around async cancellation. In short, assume you encounter something like this:

```rust
loop {
    select!(
        Some(bundle) = std::future::ready(the_factory.pop()) => {}
        Some(tx) = a_channel.recv() => {}
    }
}
```
Here, the futures are constructed upon entering the `select!` statement, but they are not polled in this but in a random order. So the ready-future is constructed by popping a bundle, moving it out of the factory and into the future. But if the channel is polled and immediately ready, then the ready-future is dropped (cancelled), and you lose the bundle.

Making the removal of the bundle lazy works around this async cancellation.